File: app/src/main/java/io/github/rainbyte/nanoxrbunits/MainActivity.java
package io.github.rainbyte.nanoxrbunits;

import android.app.Activity;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.os.Bundle;
import android.text.Editable;
import android.text.InputFilter;
import android.text.TextWatcher;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.Toast;

import java.util.Locale;
import java.util.function.Consumer;

public class MainActivity extends Activity {

    public static final String DONATION_ADDRESS = "nano_1tmt1n4is9cks3xi3eroobjosezj96aaden1pnmdngnqy7kb5nhn9koiy58s";

    private EditText NanoEdit;
    private EditText mNanoEdit;
    private EditText uNanoEdit;
    private EditText fNanoEdit;
    private EditText rawEdit;

    private static void addAfterTextChanged(EditText edit, Consumer<Editable> afterTextChanged) {
        edit.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) { }
            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) { }
            @Override
            public void afterTextChanged(Editable editable) {
                afterTextChanged.accept(editable);
            }
        });
    }

    private static InputFilter inputFilter(NanoUnit unit) {
        String partL = String.format(Locale.ENGLISH, "\\d{0,%d}", unit.getIntegers());
        String partR = unit.getDecimals() > 0
                ? String.format(Locale.ENGLISH, "(\\.\\d{0,%d})?", unit.getDecimals())
                : "";
        String pattern = partL + partR;

        return (source, start, end, dest, dstart, dend) -> {
            StringBuilder builder = new StringBuilder(dest);
            builder.replace(dstart, dend, source.subSequence(start, end).toString());
            while (builder.length() > 0 && builder.charAt(0) == '0') {
                builder.deleteCharAt(0);
            }
            return builder.toString().matches(pattern) ? null : "";
        };
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        NanoEdit = findViewById(R.id.NanoEdit);
        mNanoEdit = findViewById(R.id.mNanoEdit);
        uNanoEdit = findViewById(R.id.uNanoEdit);
        fNanoEdit = findViewById(R.id.fNanoEdit);
        rawEdit = findViewById(R.id.rawEdit);

        ImageButton qrcodeButton = findViewById(R.id.qrcodeButton);
        qrcodeButton.setOnClickListener(v -> {
            Context baseContext = getBaseContext();
            CharSequence clipboardMsgDonation = baseContext.getText(R.string.clipboard_donation);
            CharSequence clipboardMsgCopied = baseContext.getText(R.string.clipboard_copied);
            ClipboardManager clipboard = (ClipboardManager) getSystemService(Context.CLIPBOARD_SERVICE);
            if (clipboard != null) {
                ClipData clip = ClipData.newPlainText(clipboardMsgDonation, DONATION_ADDRESS);
                clipboard.setPrimaryClip(clip);
                Toast.makeText(baseContext, clipboardMsgCopied, Toast.LENGTH_SHORT).show();
            }
        });

        NanoEdit.setFilters(new InputFilter[] { inputFilter(NanoUnit.Nano) });
        mNanoEdit.setFilters(new InputFilter[] { inputFilter(NanoUnit.mNano) });
        uNanoEdit.setFilters(new InputFilter[] { inputFilter(NanoUnit.uNano) });
        fNanoEdit.setFilters(new InputFilter[] { inputFilter(NanoUnit.fNano) });
        rawEdit.setFilters(new InputFilter[] { inputFilter(NanoUnit.Raw) });

        addAfterTextChanged(NanoEdit, it -> {
            if (NanoEdit.hasFocus()) {
                update(NanoUnit.Nano, it.toString());
            }
        });
        addAfterTextChanged(mNanoEdit, it -> {
            if (mNanoEdit.hasFocus()) {
                update(NanoUnit.mNano, it.toString());
            }
        });
        addAfterTextChanged(uNanoEdit, it -> {
            if (uNanoEdit.hasFocus()) {
                update(NanoUnit.uNano, it.toString());
            }
        });
        addAfterTextChanged(fNanoEdit, it -> {
            if (fNanoEdit.hasFocus()) {
                update(NanoUnit.fNano, it.toString());
            }
        });
        addAfterTextChanged(rawEdit, it -> {
            if (rawEdit.hasFocus()) {
                update(NanoUnit.Raw, it.toString());
            }
        });

        update(NanoUnit.Nano, "1", false);
    }

    private void update(NanoUnit baseUnit, String input, boolean omitBaseUnit) {
        String raws = baseUnit.toRaws(input);
        for (NanoUnit unit : NanoUnit.values()) {
            if (unit == baseUnit && omitBaseUnit) continue;
            switch (unit) {
                case Nano:
                    NanoEdit.setText(NanoUnit.Nano.fromRaws(raws));
                    break;
                case mNano:
                    mNanoEdit.setText(NanoUnit.mNano.fromRaws(raws));
                    break;
                case uNano:
                    uNanoEdit.setText(NanoUnit.uNano.fromRaws(raws));
                    break;
                case fNano:
                    fNanoEdit.setText(NanoUnit.fNano.fromRaws(raws));
                    break;
                case Raw:
                    rawEdit.setText(NanoUnit.Raw.fromRaws(raws));
                    break;
            }
        }
    }

    private void update(NanoUnit baseUnit, String input) {
        update(baseUnit, input, true);
    }
}

File: app/src/main/java/io/github/rainbyte/nanoxrbunits/MainActivity.kt
package io.github.rainbyte.nanoxrbunits

import android.app.Activity
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.os.Bundle
import android.text.Editable
import android.text.InputFilter
import android.text.TextWatcher
import android.widget.EditText
import android.widget.Toast
import kotlinx.android.synthetic.main.activity_main.*

const val DONATION_ADDRESS = "nano_1tmt1n4is9cks3xi3eroobjosezj96aaden1pnmdngnqy7kb5nhn9koiy58s"

fun EditText.doAfterTextChanged(afterTextChanged: (Editable?) -> Unit) {
    this.addTextChangedListener(object : TextWatcher {
        override fun beforeTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {
        }
        override fun onTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {
        }
        override fun afterTextChanged(editable: Editable?) {
            afterTextChanged.invoke(editable)
        }
    })
}

class MainActivity : Activity() {

    private val NanoUnit.pattern: String
        get() = """\d{0,${integers}}""" + if (decimals > 0) """(\.\d{0,${decimals}})?""" else ""

    private fun NanoUnit.inputFilter() = InputFilter { source, start, end, dest, dstart, dend ->
        val regex = Regex(pattern)
        val builder = StringBuilder(dest)
        builder.replace(dstart, dend, source.subSequence(start, end).toString())
        val result = builder.dropWhile { it == '0' }
        if (source != null && regex.matches(result)) null else ""
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        qrcodeButton.setOnClickListener {
            val clipboardMsgDonation = baseContext.getText(R.string.clipboard_donation)
            val clipboardMsgCopied = baseContext.getText(R.string.clipboard_copied)
            val clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
            val clip = ClipData.newPlainText(clipboardMsgDonation, DONATION_ADDRESS)
            clipboard.setPrimaryClip(clip)
            Toast.makeText(baseContext, clipboardMsgCopied, Toast.LENGTH_SHORT).show()
        }

        NanoEdit.filters = arrayOf(NanoUnit.Nano.inputFilter())
        mNanoEdit.filters = arrayOf(NanoUnit.mNano.inputFilter())
        uNanoEdit.filters = arrayOf(NanoUnit.uNano.inputFilter())
        rawEdit.filters = arrayOf(NanoUnit.Raw.inputFilter())

        NanoEdit.doAfterTextChanged {
            if (NanoEdit.hasFocus()) {
                update(NanoUnit.Nano, it.toString())
            }
        }
        mNanoEdit.doAfterTextChanged {
            if (mNanoEdit.hasFocus()) {
                update(NanoUnit.mNano, it.toString())
            }
        }
        uNanoEdit.doAfterTextChanged {
            if (uNanoEdit.hasFocus()) {
                update(NanoUnit.uNano, it.toString())
            }
        }
        fNanoEdit.doAfterTextChanged {
            if (fNanoEdit.hasFocus()) {
                update(NanoUnit.fNano, it.toString())
            }
        }
        rawEdit.doAfterTextChanged {
            if (rawEdit.hasFocus()) {
                update(NanoUnit.Raw, it.toString())
            }
        }

        update(NanoUnit.Nano, "1", false)
    }

    fun update(baseUnit: NanoUnit, input: String, omitBaseUnit: Boolean = true) {
        val raws = baseUnit.toRaws(input)
        for (unit in NanoUnit.values()) {
            if (unit == baseUnit && omitBaseUnit) continue
            when (unit) {
                NanoUnit.Nano -> NanoEdit.setText(NanoUnit.Nano.fromRaws(raws))
                NanoUnit.mNano -> mNanoEdit.setText(NanoUnit.mNano.fromRaws(raws))
                NanoUnit.uNano -> uNanoEdit.setText(NanoUnit.uNano.fromRaws(raws))
                NanoUnit.fNano -> fNanoEdit.setText(NanoUnit.fNano.fromRaws(raws))
                NanoUnit.Raw -> rawEdit.setText(NanoUnit.Raw.fromRaws(raws))
            }
        }
    }
}
